### ISSUE TITLE
Use removeWhiteSpace in 922110

### ISSUE THREAD

**Issue by theseion** [coreruleset/coreruleset#3114]
As detailed in #3046, there's a bug in libmodsecurity3 where line break characters may be part of the header value, e.g. `charset=utf-8\r\n`. This leads to false positives in rule 922110. @airween found that using `t:removeWhiteSpace` fixes the false positives.

It appears to me that it should be possible to simplify the regular expression in 922110 by adding `t:removeWhiteSpace`, without impacting detection quality.

**Comment by airween** [comment]
> It appears to me that it should be possible to simplify the regular expression in 922110 by adding `t:removeWhiteSpace`, without impacting detection quality.

Sorry, I don't understand this: should we simplify the regex or use the mentioned transformation? Or how can we simplify the regex with the transformation?

**Comment by airween** [comment]
> My question about this one is: is there another place where this bug might affect rules?

I assume yes, but we have to find some way to check.

Eg. - and I'm curious your opinion - turn the debug.log, start a full regression test, and look the `\r\n` sequences  in the log. Libmodsecurity3 writes the pattern into the log.

**Comment by theseion** [comment]
> Sorry, I don't understand this: should we simplify the regex or use the mentioned transformation? Or how can we simplify the regex with the transformation?

Using `t:removeWhiteSpace` means we can remove all whitespace matching tokens from the regex, mainly `\s*`.

**Comment by airween** [comment]
> Using `t:removeWhiteSpace` means we can remove all whitespace matching tokens from the regex, mainly `\s*`.

Oh, got it, thanks.

**Comment by theseion** [comment]
> My question about this one is: is there another place where this bug might affect rules?

I was wondering the same thing. Since it is only related to multipart requests we can certainly narrow that down. I think only 3 rules should be affected, the ones that look at `MULTI_PART_HEADERS`: 922100, 922110, 922120.

- 922100 is affected I think, as it would check that `utf-8\r\n` is in `|utf-8| |iso-8859-1| |iso-8859-15| |windows-1252|`
- 922110 is the rule we're talking about here
- 922120 is not affected IMO, as it  uses an unconditional match (`.*`)

BTW, I think there's a bug in 922100. `allowed_request_content_type_charset` is `|utf-8| |iso-8859-1| |iso-8859-15| |windows-1252|. If `ARGS:_charset_` were `iso-8859-1`, it would match both `iso-8859-1` and `iso-8859-15`. The correct way to use `within` here would be to use `setvar` and remove the superfluous vertical bars:

```
SecRule &MULTIPART_PART_HEADERS:_charset_ "!@eq 0" \
    "id:922100,\
    phase:2,\
    block,\
    t:none,\
    msg:'Multipart content type global _charset_ definition is not allowed by policy',\
    logdata:'Matched Data: %{ARGS._charset_}',\
    tag:'application-multi',\
    tag:'language-multi',\
    tag:'platform-multi',\
    tag:'attack-multipart-header',\
    tag:'OWASP_CRS',\
    tag:'capec/1000/255/153',\
    tag:'paranoia-level/1',\
    ver:'OWASP_CRS/4.0.0-rc1',\
    severity:'CRITICAL',\
    chain"
    SecRule ARGS:_charset_ "@rx ^.*$" \
        "setvar:'tx.charset=|%{tx.0}|',\
          chain"
          SecRule TX:charset "!@within %{tx.allowed_request_content_type_charset}" \
              "t:lowercase,\
              setvar:'tx.inbound_anomaly_score_pl1=+%{tx.critical_anomaly_score}'"
```

Do you agree?

**Comment by airween** [comment]
>```
>    severity:'CRITICAL',\
>    chain"
>    SecRule ARGS:_charset_ "@rx ^.*$" \
>        "setvar:'tx.charset=|%{tx.0}|',\
>        chain"
>        SecRule TX:charset "!@within %{tx.allowed_request_content_type_charset}" \
>            "t:lowercase,\
>```

May be a `capture` is missing from here? And then the `setvar` is no necessary:
```
    severity:'CRITICAL',\
    chain"
    SecRule ARGS:_charset_ "@rx ^.*$" \
        "capture,\
        chain"
        SecRule TX:0 "!@within %{tx.allowed_request_content_type_charset}" \
             "t:lowercase,\
             setvar:'tx.inbound_anomaly_score_pl1=+%{tx.critical_anomaly_score}'"
```

(I haven't checked...)

**Comment by theseion** [comment]
No, `tx.0` is always the full match of the expression, no `capture` required. `setvar` is necessary, because I'm adding vertical bars around the match: `|%{tx.0}|`. That's the gist of the trick.

**Comment by fzipi** [comment]
Hey guys, what's next here?

**Comment by theseion** [comment]
We need to create issues / PRs for the affected rules (922100, 922120).

**Comment by theseion** [comment]
https://github.com/owasp-modsecurity/ModSecurity/pull/2870 has been merged. The solution here would be for legacy support. Note that it might be difficult / impossible to apply the idea to 922100.

**Comment by fzipi** [comment]
Is there any follow-up here?

**Comment by theseion** [comment]
There should be, but I never got around to it. Too many other things to do :)

**Comment by fzipi** [comment]
So #3470 fixed this for 922100. We need one for 922120.

**Comment by fzipi** [comment]
@theseion I don't think this needs to be applied to 922120 🤔 Maybe you meant other rule?

**Comment by theseion** [comment]
You're right. 922120 would be affected but doesn't include whitespace in the regex:
> 922120 is not affected IMO, as it uses an unconditional match (.*)

**Comment by theseion** [comment]
Actually, I don't believe that #3470 fixed the issue we're talking about here. If the value of `_charset_` is `utf-8\r\n`, it will not match (which would actually lead to the rule blocking a request when it shouldn't). What @azurit fixed was that `@within` was being used wrongly.

**Comment by azurit** [comment]
I think there is a typo in the PR description - first, i'm mentioning issue 3114 (this) but at the bottom, the issue 3118 (which seems to be the right one).

**Comment by theseion** [comment]
Thanks @azurit, I've updated the other PR to point to #3118 in both cases.

**Comment by fzipi** [comment]
@theseion @azurit Honestly, this was fixed upstream in Oct 2024. I don't think we should do anything here.

Let's close and move on. We can document the need for latest libmodsecurityv3 if needed.

**Comment by azurit** [comment]
Looks like the original issue (in libmodsecurity) was fixed in version [3.0.10](https://github.com/owasp-modsecurity/ModSecurity/releases/tag/v3.0.10) release in July 2023. We should make version 3.0.10 as recommended minimum ale close this.
